### DISCUSSION TITLE
Allow setting the name of SSL certificate header (currently `SSL_CLIENT_CERT`) when EJBCA behind proxy

### DISCUSSION THREAD

**Discussion by akielbas** [Keyfactor/ejbca-ce#655] [Ideas]
Is `SSL_CLIENT_CERT` hardcoded somewhere in EJBCA? Is it possible to make it configurable?

I'm interested in perhaps running EJBCA behind AWS Application Load Balancer which supports mTLS but passes the client certificate as `X-Amzn-Mtls-Clientcert` (per [documentation](https://docs.aws.amazon.com/elasticloadbalancing/latest/application/mutual-authentication.html#mtls-http-headers)).

EDIT: After submitting this I found [this thread](https://sourceforge.net/p/ejbca/discussion/123123/thread/b37f58cf33/) on the old sourceforge site. Perhaps something has changed since then? Would rather avoid running an intermediate proxy to rewrite header name.

**Comment by primetomas** [comment]
Nothing has changed. EJBCA itself does not code any dependency on headers.
To my knowledge, EJBCA SaaS uses AWS load balancer so somehow this is possible.
https://docs.keyfactor.com/ejbca-saas/latest/aws/

**Comment by akielbas** [comment]
Looks like this might be possible now from ELB side:
https://docs.aws.amazon.com/elasticloadbalancing/latest/application/header-modification.html

**Comment by akielbas** [comment]
I spoke too soon, I think.

I am able to get the right header name, but I'm unable to get it to work directly because the ELB header is url encoded which WildFly does not seem to support.

Reproduced this with nginx:
* Works: Using [`$ssl_client_cert` variable](https://nginx.org/en/docs/http/ngx_http_ssl_module.html#var_ssl_client_cert). However docs note that the existing variable is deprecated and we should use `$ssl_client_escaped_cert` (urlencoded version).
* Fails: Using `$ssl_client_escaped_cert`.

I've opened a [WildFly enhancement](https://issues.redhat.com/browse/WFLY-20206).

EDIT:
In case anyone else tries to get this to work.

I did end up having a small nginx container rewrite the header name and the contents with perl module (had to decode URL encoding and also add spaces on each line for the old multiline HTTP header support). Here is the `nginx.conf` hack:
```
load_module /etc/nginx/modules/ngx_http_perl_module.so;

worker_processes  auto;

events {
    worker_connections  1024;
}

http {
    perl_set $clientcert '
    sub {
        my $r = shift;
        my $in_cert = $r->header_in("X-Amzn-Mtls-Clientcert");

        # urldecode
        $in_cert =~ s/%([0-9A-Fa-f]{2})/chr(hex($1))/eg;

        # indent all the lines except the first
        $in_cert =~ s/^(?!.*\A).*/    $&/gm;

        # remove trailing newline
        $in_cert =~ s/\n$//; 

        return $in_cert;
    }
    ';

    server {
        listen 8082;
        location / {
            proxy_set_header SSL_CLIENT_CERT $clientcert;
            proxy_set_header X-Amzn-Mtls-Clientcert '';
            proxy_set_header Host $http_host;
            proxy_pass http://ejbca:8082;            
        }
    }
}
```

I used it with `nginx:alpine-perl` container. If someone has a more elegant way of doing this I'm open to suggestions.

**Comment by luckylinux** [comment]
In my Case (Homelab, definitively not behind AWS) using the deprecated `proxy_set_header SSL_CLIENT_CERT $ssl_client_cert;` seems to be the only Thing that makes it work. Finally after 3 Days and trying to get EJBCA to work first behind Caddy, then Traefik and finally NGINX !

Alternatively I guess using @akielbas Approach I guess we would have to get the `$ssl_client_escaped_cert`, then url-decode it, and use that instead.

A shame that NGINX decided to deprecate that Variable. It works, not sure until when, but not sure when it will be eventually removed :confused:.
